crates/interpreter/src/valid.rs

```diff
@@ -25,23 +25,47 @@ use crate::toctou::*;
 use crate::util::*;
 use crate::*;
 
+pub trait ValidMode {}
+
+pub struct Prepare;
+impl ValidMode for Prepare {}
+
+impl Default for Prepare {
+    fn default() -> Self {
+        Self
+    }
+}
+
 /// Checks whether a WASM module in binary format is valid.
-pub fn validate(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
+pub fn validate<M: ValidMode + Default>(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
     Context::default().check_module(&mut Parser::new(binary))
 }
 
 type Parser<'m> = parser::Parser<'m, Check>;
 type CheckResult = MResult<(), Check>;
 
-#[derive(Default)]
-struct Context<'m> {
+struct Context<'m, M: ValidMode = Prepare> {
```

Comment:
Do we need the default? Can't we just have `M: ValidMode`? Ideally we should always be explicit.